// react-inner-image-zoom/lib/components/Image.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
var Image = function Image(_ref) {
  var src = _ref.src,
    sources = _ref.sources,
    width = _ref.width,
    height = _ref.height,
    hasSpacer = _ref.hasSpacer,
    imgAttributes = _ref.imgAttributes,
    isZoomed = _ref.isZoomed,
    fadeDuration = _ref.fadeDuration;
  var createSpacer = width && height && hasSpacer;
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      paddingTop: createSpacer ? "".concat(height / width * 100, "%") : null
    }
  }, sources && sources.length > 0 ? /*#__PURE__*/_react.default.createElement("picture", null, sources.map(function (source, i) {
    return /*#__PURE__*/_react.default.createElement(_react.Fragment, {
      key: i
    }, source.srcSet && /*#__PURE__*/_react.default.createElement("source", source));
  }), /*#__PURE__*/_react.default.createElement("img", _extends({}, imgAttributes, {
    className: "iiz__img ".concat(imgAttributes.className || '', " ").concat(isZoomed ? 'iiz__img--hidden' : '', " ").concat(createSpacer ? 'iiz__img--abs' : ''),
    style: {
      transition: "opacity 0ms linear ".concat(isZoomed ? fadeDuration : 0, "ms, visibility 0ms linear ").concat(isZoomed ? fadeDuration : 0, "ms")
    },
    src: src,
    width: width,
    height: height
  }))) : /*#__PURE__*/_react.default.createElement("img", _extends({}, imgAttributes, {
    className: "iiz__img ".concat(imgAttributes.className || '', " ").concat(isZoomed ? 'iiz__img--hidden' : '', " ").concat(createSpacer ? 'iiz__img--abs' : ''),
    style: {
      transition: "opacity 0ms linear ".concat(isZoomed ? fadeDuration : 0, "ms, visibility 0ms linear ").concat(isZoomed ? fadeDuration : 0, "ms")
    },
    src: src,
    width: width,
    height: height
  })));
};
var _default = Image;
exports.default = _default;
